### PR TITLE
 Initial config implementation

### PR DESCRIPTION
**Pull request by PereBal**

### CODE REVIEW

File: bigrig/config.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        try:
+            with open(path, "rt") as fid:
+                credentials = yaml.safe_load(fid)
+        except yaml.YAMLError:
+            # Raise from None to avoid accidental leaks
+            raise yaml.YAMLError(
+                f"Unable to load credentials from `{path}`, it appears to be a non-yaml file"
+            ) from None
+        except OSError as exc:
+            raise OSError(f"Unable to load credentials from `{path}`") from exc
```
Comment (thule0) [inline bigrig/config.py]: Nitpick: is there a lot of point of catching and reraising the same error with lightly different message?

File: bigrig/config.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    @classmethod
+    def from_dict(cls, blob: t.Dict) -> "Origin":
+        return cls(
+            location=blob["location"],
+            credentials=OriginCredentials.from_path(path=blob.get("credentialsPath")),
+        )
+
+
+@dataclass
+class OriginCredentials:
```
Comment (thule0) [inline bigrig/config.py]: I see only origin seems to have credentials, and naming reflects that.
All the repos should have it actually. In fact, the way it works right now is  you need password only to upload, so origin is the only one that does not require credentials :)

File: bigrig/config.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    @classmethod
+    def from_dict(cls, blob: t.Dict) -> "Target":
+        return cls(location=blob["location"], variables=blob["variables"])
+
+
+@dataclass
+class ConfigEntry:
+    origin: Origin
+    source: Source
+    target: t.Dict[str, Target]
```
Comment (thule0) [inline bigrig/config.py]: targetS?

**Comment by thule0** [comment]
Can we have some tests to see what is valid?  JsonSchema is a bit of a mouthful tbh.

File: bigrig/config.py (hunk 2 above)
Comment (PereBal) [reply bigrig/config.py]: This is what we discussed initially, but I think it just makes sense to allow having credentials in every origin,source,target entry, I'll review and update the design doc accordingly :)

File: bigrig/config.py (hunk 1 above)
Comment (PereBal) [reply bigrig/config.py]: As discussed via slack, I tend to catch these kind of exceptions to provide more user-friendly messages, while I agree that this one in particular is just the same thing, I'd like to leave it as is for consistency with the rest of the module.

**Comment by PereBal** [comment]
> Can we have some tests to see what is valid? JsonSchema is a bit of a mouthful tbh.

@roman-alexeev What would you like to cover?

File: bigrig/config.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        if not packages_path:
+            raise ValueError(
+                f"Missing path to bigrig package list, point BIGRIG_PACKAGES_PATH to bigrig"
+                f"package file list"
+            )
+
+        try:
+            with open(config_path, "rt") as fid:
+                entries = [
+                    ConfigEntry.from_dict(blob=blob) for blob in yaml.safe_load_all(fid)
```
Comment (thule0) [inline bigrig/config.py]: We can have more then one?

Comment (PereBal) [reply bigrig/config.py]: As you proposed, yes. And I would like to point out that allowing for more than one config object at a time, lets us have multiple origins to multiple targets, etc in a fairly trivial way so I'd leave it as is.

Comment (thule0) [reply bigrig/config.py]: Did I? :)
So you mean going through configs as a list and run for each of them?

**Review by thule0 — APPROVED** [review]

File: bigrig/config.py (hunk 4 above)
Comment (PereBal) [reply bigrig/config.py]: yea, this is related to my questions regarding packages per origin and what not :)

**Review by thule0 — APPROVED** [review]

File: bigrig/config.py (hunk 4 above)
Comment (PereBal) [reply bigrig/config.py]: removed as agreed on slack, multiple origin/target will be handled by multiple instances of the app ( for now )

**Pull request merged**